src/common/src/mlib/str.h

```diff
@@ -565,4 +642,488 @@ mstr_contains_any_of(mstr_view str, mstr_view needle)
+   mlib_check(!mlib_sub(&new_len, n_delete));
+   // Check if appending would make too big of a string
+   if (mlib_add(&new_len, insert.len)) {
+      // New string will be too long
+      return false;
+   }
+   char *mut = str->data;
+   // We either resize first or resize last, depending on where we are shifting chars
+   if (new_len > str->len) {
+      // Do the resize first
+      if (!mstr_resize_for_overwrite(str, new_len)) {
+         // Failed to allocate
+         return false;
+      }
+      mut = str->data;
+   }
+   // Move to the splice position
+   mut += splice_pos;
+   // Shift the existing string parts around for the deletion operation
+   const size_t tail_len = n_chars_avail_to_delete - n_delete;
+   // Adjust to the begining of the string part that we want to keep
+   char *copy_from = mut + n_delete;
+   char *copy_to = mut + insert.len;
+   memmove(copy_to, copy_from, tail_len);
+   if (new_len < str->len) {
+      // We didn't resize first, so resize now. We are shrinking the string, so this
+      // will never fail, and does not create any uninitialized memory:
+      mlib_check(mstr_resize_for_overwrite(str, new_len));
+      mut = str->data + splice_pos;
+   }
+   // Insert the new data
+   memcpy(mut, insert.data, insert.len);
+   return true;
+}
+
+/**
+ * @brief Append a string to the end of some other string.
+ *
+ * @param str The string to be modified
+ * @param suffix The suffix string to be appended onto `*str`
+ * @return true If the operation was successful
+ * @return false Otherwise
+ *
+ * If case of failure, `*str` is not modified.
+ */
+static inline bool
+mstr_append(mstr *str, mstr_view suffix)
+{
+   return mstr_splice(str, str->len, 0, suffix);
+}
+
+#define mstr_append(Into, Suffix) mstr_append((Into), mstr_view_from((Suffix)))
+
+/**
+ * @brief Append a single character to the given string object
+ *
+ * @param str The string object to be updated
+ * @param c The single character that will be inserted at the end
+ * @return true If the operation succeeded
+ * @return false Otherwise
+ *
+ * In case of failure, the string is not modified.
+ */
+static inline bool
+mstr_pushchar(mstr *str, char c)
+{
+   mstr_view one = mstr_view_data(&c, 1);
+   return mstr_append(str, one);
+}
+
+/**
+ * @brief Replace every occurrence of `needle` in `str` with `sub`
+ *
+ * @param str The string object to be updated
+ * @param needle The non-empty needle string to be searched for.s
+ * @param sub The string to be inserted in place of each `needle`
+ * @return true If the operation succeeds
+ * @return false Otherwise
+ *
+ * @warning If the `needle` string is empty, then the program will terminate!
+ * @note If the operation fails, the content of `str` is an unspecified but valid
+ * string.
+ */
+static inline bool
+mstr_replace(mstr *str, mstr_view needle, mstr_view sub)
+{
+   mlib_check(needle.len, neq, 0, because, "Trying to replace an empty string will result in an infinite loop");
```

Comment:
Perhaps no-op may be preferable to termination here, so users need not special-case an empty `needle`?